test/csuite/wt3363_checkpoint_op_races/main.c

```diff
@@ -0,0 +1,498 @@
+ * checkpoint and measure that no operation takes 1/2 the length of this delay.
+ *
+ * Failure mode: We monitor the execution time of all operations and if we find
+ * any operation taking longer than 1/2 the delay time, we abort dumping a core
+ * file which can be used to determine what operation was blocked.
+ */
+
+int handle_error(WT_EVENT_HANDLER *, WT_SESSION *, int, const char *);
+int handle_message(WT_EVENT_HANDLER *, WT_SESSION *, const char *);
+void* monitor(void *);
+void* do_checkpoints(void *);
+void* do_ops(void *);
+void op_bulk(WT_CONNECTION *);
+void op_bulk_unique(WT_CONNECTION *, int, int);
+void op_create(WT_CONNECTION *);
+void op_create_unique(WT_CONNECTION *, int, int);
+void op_cursor(WT_CONNECTION *);
+void op_drop(WT_CONNECTION *, int);
+
+#define	N_THREADS 10
+#define	RUNTIME 900.0
+
+typedef struct {
+	TEST_OPTS *testopts;
+	int threadnum;
+	int nthread;
+	int done;
+} THREAD_ARGS;
+
+static pthread_rwlock_t single;
+/*
+ * Time delay to introduce into checkpoints in seconds. Should be at-least
+ * double the maximum time that any one of the operations should take. Currently
+ * this is set to 10 seconds and we expect no single operation to take longer
+ * than 5 seconds.
+ */
+static const uint64_t max_execution_time = 10;
+
+/* Int used to make unique table names */
+static uint64_t uid = 1;
+
+/*
+ * An array of the number of operations that each of the worker threads has
+ * performed.
+ */
+static uint64_t thread_counters[N_THREADS];
+static const char *uri;
+
+int
+main(int argc, char *argv[])
+{
+	static WT_EVENT_HANDLER event_handler = {
+		handle_error,
+		handle_message,
+		NULL,
+		NULL
+	};
+	TEST_OPTS *opts, _opts;
+	THREAD_ARGS thread_args[N_THREADS];
+	pthread_t ckpt_thread, mon_thread, threads[N_THREADS];
+	int i;
+
+	if (!testutil_enable_long_tests())	/* Ignore unless requested */
+		return (EXIT_SUCCESS);
+
+	opts = &_opts;
+	memset(opts, 0, sizeof(*opts));
+
+	testutil_check(testutil_parse_opts(argc, argv, opts));
+	testutil_make_work_dir(opts->home);
+
+	testutil_check(wiredtiger_open(
+	    opts->home, &event_handler, "create,cache_size=1G,", &opts->conn));
+
+	uri = opts->uri;
+
+	testutil_check(pthread_create(
+	    &ckpt_thread, NULL, do_checkpoints, (void *)opts->conn));
+
+	for (i = 0; i < N_THREADS; ++i) {
+		thread_counters[i] = 0;
+		thread_args[i].threadnum = i;
+		thread_args[i].nthread = N_THREADS;
+		thread_args[i].testopts = opts;
+		testutil_check(pthread_create(
+		    &threads[i], NULL, do_ops, (void *)&thread_args[i]));
+	}
+
+	testutil_check(pthread_create(&mon_thread, NULL, monitor, NULL));
+
+	for (i = 0; i < N_THREADS; ++i)
+		testutil_check(pthread_join(threads[i], NULL));
+
+	testutil_check(pthread_join(mon_thread, NULL));
+
+	testutil_check(pthread_join(ckpt_thread, NULL));
+
+	printf("Success\n");
+
+	testutil_cleanup(opts);
+	return (EXIT_SUCCESS);
+}
+
+/* WiredTiger error handling function */
```

Comment:
It's good that you added comments for most of the functions, but it's common practice in WiredTiger to use block-style (i.e: multi-line) comments at the function level.